### DISCUSSION TITLE
What sets the initial system timestamp on first boot?

### DISCUSSION THREAD

**Discussion by wico-silva** [OE4T/meta-tegra#1694] [Q&A]
I naively assumed that on first boot the system will use the timestamp in `/etc/timestamp`. Turns out that is not true, in `/etc/timestmap` I have `20180309123456` and on first boot the system time is actually `20220428174228`.

I've done a search over all the files, looking for that timestamp and couldn't find anything.

Hence the question, what is setting the initial timestamp and how can I modify it?

Thanks!

EDIT: This question is towards a system that is not using NTP and is not connected to the internet.

**Comment by madisongh** [comment]
If you're using [systemd-timesyncd](https://www.freedesktop.org/software/systemd/man/latest/systemd-timesyncd.html#), it uses a different set of files. See the linked man page for more info.

**Reply by wico-silva** [reply]
Thanks for the reply!
Forgot to mention that I'm not using an NTP server 😅 . The system is not connected to the internet.

**Reply by madisongh** [reply]
It's more based on whether or not you're using systemd (vs. sysvinit scripts), and whether you've included systemd-timesyncd in your builds. I don't know what, if anything, systemd does with the system clock if systemd-timesyncd is _not_ included.

If you use sysvinit, there should be an init script you can look at to see the logic for setting the system clock at boot time.

Either way, you should be able to customize the behavior to suit your needs.

**Reply by wico-silva** [reply]
Understood, thanks for the explanation. I got it working with the following changes:

```
ROOTFS_POSTPROCESS_COMMAND += "set_initial_sys_time;"
set_initial_sys_time() {
    touch -md '@${REPRODUCIBLE_TIMESTAMP_ROOTFS}' ${IMAGE_ROOTFS}/usr/lib/clock-epoch
}
```

In local.conf:
```
REPRODUCIBLE_TIMESTAMP_ROOTFS = "${@bb.process.run('date +%s')[0].strip()}"
```

On first boot, the system time is now the same as the build time. Thanks @madisongh!
